### ISSUE TITLE
[Module Proposal]: AKS secret autorotation

### ISSUE THREAD

**Issue by little-evan** [Azure/bicep-registry-modules#275]
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

To provide an umbrella module that will create necessary resources so that secret in a Azure keyvault that stores e.g. a CosmosDB access key can be rotated automatically when it's about to expire.

### Module path

keyvault/secret-rotation

### Describe the module

This module creates a Function app with the necessary components such storage account, system event grid, log analytic workspace etc. The app hosts different functions that will subscribe to the SecretNearExpiry event published by the given keyvault targeting the specific secret, so once the secret is about to expire, the corresponding function will be executed and renew the original access key on the target resource for example CosmosDB, and update the keyvault secret with the generated key. The module supports integration with vnet so that the functions can access resources privately. Currently the module supports rotating CosmosDB and Redis access keys, and can be extended to other types of resource as well.

**Comment by Gordonby** [comment]
Love the idea.

Not sure if this is a better fit for the [Azure Quickstart Templates](https://github.com/Azure/azure-quickstart-templates) though. Where would the function app code live?

@alex-frankel - thoughts?

**Comment by shenglol** [comment]
> Where would the function app code live?

Maybe publish the code as a container image?

**Comment by vlahane** [comment]
Function app code would live here: https://github.com/Azure/bicep-registry-modules/issues/259 can be integrated to AKV secret autorotation (this proposed module) if use wanted to create new or use existing function app, but this proposed module will be more focused on secret autorotation and using existing function app module. 

cc_ @little-evan

**Comment by little-evan** [comment]
Thanks for sharing your thoughts folks :-)
The app code currently is within the module itself, essentially the whole function app is packaged as a zip file and loaded as base64 encoded text when compiled by bicep into the ARM templated, a Azure deployment script is used to deploy the source code to the function app. 
A bit of inconvenience is that every time you make change to the function source code, you'll have to repackage and zip file and commit. However the function source code is relatively stable I would say so we probably will not expect that much changes to them. 
We can also explore the container option @shenglol  mentioned, if that fits better.

**Comment by Gordonby** [comment]
What i've done a few times is leverage sourcecontrol binding from a public github repository.
This way the code could have it's own lifecycle in its own repo using tags/releases as the reference from the Bicep registry module.

eg. https://github.com/Gordonby/AzureBicepServerlessAppStack/blob/509cb1b27882965d26e618d99e3db722610ee894/bicep/foundation/functionapp.bicep#L128

**Comment by vlahane** [comment]
as I remembered sourcecontrols resource was unstable that why we end up using deployment script long back, @little-evan do we have error message.

**Comment by little-evan** [comment]
Yep we do experienced some hiccups with sourcecontrols before, but the reason we gave up on it is more because our resources are run on private network and we don't want to have dependency on external resources, and don't have to worry about firewall stuff. However now that this is a public module, I think it'll make sense to provide sourcecontrols as one of the options and it's up to the users to decide what to use eventually.

**Comment by little-evan** [comment]
Hello, do we have an update regarding this proposal?

**Comment by shenglol** [comment]
Sorry for the delay. Feel free to create a PR (though we are still blocked on https://github.com/Azure/bicep-registry-modules/issues/286).
